### PR TITLE
Implement Bucket Notifications APIs

### PR DESCRIPTION
**Pull request by donatello**
Towards resolving #394 

At this point it is an early preview, and it is not yet complete.

### CODE REVIEW

File: minio/api.py [hunk 1]
```diff
@@ -434,19 +434,18 @@ def remove_all_bucket_notifications(self, bucket_name):
         """
         is_valid_bucket_name(bucket_name)
 
-        content = xml_marshal_bucket_notifications({})
+        content_bytes = xml_marshal_bucket_notifications({})
         headers = {
-            'Content-Length': str(len(content)),
-            'Content-MD5': encode_to_base64(get_md5(content.encode('utf-8')))
```
Comment (harshavardhana) [inline minio/api.py]: Can you test this with 'python 3.5' - the changes related to 'utf-8' were done with this purpose?

**Comment by harshavardhana** [comment]
Can you add some unit-tests as well here located at `tests/unit`

File: minio/api.py (hunk 1 above)
Comment (donatello) [inline minio/api.py]: I am using python 3.5 by default for development. It works fine. It also works in python 2.7

**Comment by donatello** [comment]
Ok at this point APIs are implemented and do appear to work in my testing. Up for review.

Working on unit testing.

**Comment by harshavardhana** [comment]
> Ok at this point APIs are implemented and do appear to work in my testing (both python 3.5 and python 2.7). Up for review.
> 
> Working on unit testing.

Will review ..

**Comment by donatello** [comment]
Added a batch of unit tests.

File: minio/xml_marshal.py [hunk 2]
```diff
@@ -66,3 +66,134 @@ def xml_marshal_complete_multipart_upload(uploaded_parts):
...
+
+def xml_marshal_bucket_notifications(notifications):
+    """
+    Marshals the notifications structure for sending to S3 compatible storage
+
+    :param notifications: Dictionary with following structure:
+
+    {
+        'TopicConfigurations': [
+            {
```
Comment (donatello) [inline minio/xml_marshal.py]: Need to decide on an appropriate place to put the in-code documentation for the structure expected by the set_bucket_notification() API and returned by get_bucket_notification() API.

The expected structure is indicated here at this point.

**Comment by donatello** [comment]
Next up is documentation in API.md.

**Comment by donatello** [comment]
Aargh. Tests fail because Windows UA does not match the ones in the unit tests. Need to figure/discuss this.

File: docs/API.md [hunk 3]
```diff
@@ -319,6 +322,218 @@ minioClient.set_bucket_policy('mybucket',
...
+
+|Param   |Type   |Description   |
+|:---|:---|:---|
+|``bucket_name``   | _string_  |Name of the bucket.|
+
+__Return Value__
+
+|Param   |Type   |Description   |
+|:---|:---|:---|
+|``Policy``   | _dict_   | If there is no notification configuration, an empty dictionary is returned. Otherwise it has the same structure as the argument to set_bucket_notification   |
```
Comment (harshavardhana) [inline docs/API.md]: This is not Policy.. should be `NotificationConfig`

File: docs/API.md [hunk 4]
```diff
@@ -319,6 +322,218 @@ minioClient.set_bucket_policy('mybucket',
...
+notification = minioClient.get_bucket_notification('mybucket')
+# If no notification is present on the bucket:
+# notification == {}
+
+```
+
+<a name="set_bucket_notification"></a>
+### set_bucket_notification(bucket_name, notification)
+
+Fetch the notifications configuration on a bucket.
```
Comment (harshavardhana) [inline docs/API.md]: This is incorrect  `Fetch the notifications configuration on a bucket` is not the purpose here.  We need to change this to reflect what the `set_bucket_notification` should be doing.

File: minio/xml_marshal.py (hunk 2 above)
Comment (harshavardhana) [inline minio/xml_marshal.py]: This is fine.

File: docs/API.md (hunk 4 above)
Comment (donatello) [inline docs/API.md]: I was still iterating on it. Wanted to see how it looked. Forgot it would update the PR as well. Still working on it.

Comment (donatello) [inline docs/API.md]: I am going to be pushing and squashing more commits now. I will ping when its ready for review. Please wait :-)

Comment (harshavardhana) [inline docs/API.md]: Sure :+1:

**Comment by donatello** [comment]
@harshavardhana Ready for review and perhaps merge!

**Pull request merged**